### PR TITLE
feat: add balancec-watcher updater [pr 2 - connection logic]

### PR DESCRIPTION
**Pull request by limitofzero**
# Summary

Wires the cowswap-frontend to the new balances-watcher SSE service, behind the LaunchDarkly flag isBwEnabled.

When the flag is on, balances are pushed in real time over an EventSource and the existing multicall pipeline (priority tokens + full token-list polling) is bypassed. When off, nothing changes — the multicall path stays as is.

Note on the native token (ETH / xDAI / MATIC / BNB): the watcher service does not emit native balances, so we keep polling them via a single eth_getBalance RPC call (wagmi useBalance), refetched every 11 seconds.
This means the native balance can lag up to 11s behind the on-chain state — that's an explicit trade-off because I reused current implementation, but if it's crucial - I can rework it. 

Out of scope (planned for follow-up):
  - Fallback to multicall when the SSE stream terminates with an unrecoverable error.
  - if the FE tries to send more than 1500 tokens (token lists tokens + custom tokens) - it returns the error. Because it's the only second pr - I decided to not implement fallback on it, it will be implemented further when we decide what to do [here](https://nomevlabs.slack.com/archives/C0361CDG8GP/p1781705435600939)) 

# To Test

1. Open the swap widget and switch on isBwEnabled flag. 

- [ ] Network tab shows a single POST /sessions followed by an open EventSource to the balances-watcher endpoint - and it returns 200 (OK)
- [ ] Balances of imported tokens / tokens from enabled lists are visible
- [ ] try to swap - balances should be updated very fast, about 0-2s after order execution
- [ ] if user switch on token list/or add custom token - there should be another  /sessions call with new token list or custom token in parameters(body), the balance of these tokens should be updated too

## Release Notes

* **New Features**
  * Added a real-time balances watcher updater to keep wallet balances in sync via live balance events.
  * Added an updater to write the connected wallet’s native token balance into state.
  * Added hooks to gather enabled token list URLs and derive per-chain custom tokens.
* **Bug Fixes**
  * Improved watcher session start/stop logic, including safer handling for empty token sources, non-EVM chains, and race conditions.
* **Tests**
  * Added comprehensive coverage for watcher session lifecycle, event merging/error handling, and native/custom token behavior.
* **Chores**
  * Updated public exports to make the new watcher updater available.

### CODE REVIEW

File: libs/balances-and-allowances/src/updaters/BalancesWatcherUpdater.tsx [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+ * Watcher-mode peer of `BalancesAndAllowancesUpdater`. Drives `balancesAtom`
+ * via the balances-watcher SSE stream and mounts the same reset/cache
+ * subtrees so the atom lifecycle (chain/account reset, localStorage
+ * hydration/persistence) stays intact when the LD flag is on.
+ */
+export function BalancesWatcherUpdater({ account, chainId }: BalancesWatcherUpdaterProps): ReactNode {
+  const tokensListsUrls = useEnabledTokensListsUrls()
+  const customTokens = useCustomTokensForChain(chainId)
+
+  useBalancesWatcherSession({ account, chainId, tokensListsUrls, customTokens })
```
Comment (Danziger) [inline libs/balances-and-allowances/src/updaters/BalancesWatcherUpdater.tsx:30]: This might be a good opportunity to implement this using only atoms instead of creating new updaters, as we discussed in the past we'd like to decouple business logic and app state from UI / React.

Comment (limitofzero) [reply libs/balances-and-allowances/src/updaters/BalancesWatcherUpdater.tsx:30]: not really get your point, is still stores data in atoms but we need to run hooks somewhere for fetching data

File: libs/balances-and-allowances/src/updaters/NativeTokenBalanceUpdater.tsx [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+}
+
+/**
+ * Writes the connected wallet's native token balance into `balancesAtom`.
+ * The balances-watcher SSE stream does not emit native amounts, so we keep
+ * polling via wagmi's `eth_getBalance` (single RPC call, no multicall).
+ */
+export function NativeTokenBalanceUpdater({ account, chainId }: NativeTokenBalanceUpdaterProps): null {
+  const updateTokenBalance = useUpdateTokenBalance()
+  const { data: nativeTokenBalance } = useNativeTokenBalance(account, chainId)
```
Comment (Danziger) [inline libs/balances-and-allowances/src/updaters/NativeTokenBalanceUpdater.tsx:21]: Same for this one.

File: libs/balances-and-allowances/src/updaters/BalancesWatcherUpdater.tsx (hunk 1 above)
Comment (Danziger) [reply libs/balances-and-allowances/src/updaters/BalancesWatcherUpdater.tsx:30]: Not really, you would do something like this to have the whole flow run without hooks:



    export const balancesAtom= atom<...>({})

    balancesAtom.onMount = () => {
      return observe((get, set) => {
        const { chainId, account } = get(walletInfoAtom)
        const tokensListsUrls = get(...)
        const customTokens = get(...)

        let cancelled = false
        let subscription: BalancesSubscription | undefined
        let isFirstEvent = true

        createBalancesWatcherSession({
          chainId,
          owner: account,
          body: { tokensListsUrls, customTokens },
        })
          .then(() => {
            if (cancelled) return
      
            subscription = subscribeToBalancesEvents({
              chainId,
              owner: account,
              onBalances: (balances) => {
                if (cancelled) return
                set(balancesAtom, (state) => writeBalancesUpdate(state, balances, chainId, isFirstEvent))
                isFirstEvent = false
              },
              onError: (error, terminal) => {
                if (cancelled) return
                // Non-terminal errors mean EventSource is reconnecting — the
                // transport recovers on its own. Only surface terminal errors.
                if (!terminal) return
                set(balancesAtom, (state) => ({ ...state, error: error.message, isLoading: false }))
              },
            })
          })
          .catch((error: unknown) => {
            if (cancelled) return
            const message = error instanceof Error ? error.message : String(error)
            set(balancesAtom, (state) => ({ ...state, error: message, isLoading: false }))
          })

        return () => {
          cancelled = true
          subscription?.close()
        }
      }, jotaiStore)
    }

**Review by Danziger — APPROVED** [review]
Approving as the comments are not blockers.

**Review by fairlighteth — COMMENTED** [review]
<details>
<summary>⚠️ <strong>AI Review</strong> (Codex GPT-5, worked 5m): watcher failures can leave swaps stuck on balance loading</summary>

**Finding:** Handle watcher errors before the first snapshot as a completed balance load failure

- Location: `libs/balances-and-allowances/src/hooks/useBalancesWatcherSession.ts:75` and `apps/cowswap-frontend/src/modules/tradeFormValidation/hooks/useTradeFormValidationContext.ts:111`
- When `isBwEnabled` is true, the PR bypasses the RPC updater and relies on the watcher path for EVM chains.
- If `createBalancesWatcherSession` rejects or the SSE stream terminates before the first `balance_update`, the hook sets `error` and `isLoading: false`, but leaves `hasFirstLoad: false`.
- The validation context then computes `isBalancesLoading: !hasFirstLoad || isBalancesLoading`, so the UI stays in `BalancesLoading` instead of surfacing `BalancesNotLoaded`. This matches the current Cypress failures on the PR head where swap flows never find `#do-trade-button`.

**Suggested fix**
- Make the error path stop the first-load gate, or make validation treat a balance error as not loading. For example, terminal watcher/create-session failures should transition to a state that renders `BalancesNotLoaded` rather than permanent `BalancesLoading`.
- Add coverage for both `createSession` rejection and terminal SSE error before the first snapshot, asserting the downstream validation sees a not-loaded/error state rather than loading forever.

<details>
<summary>Review scope and related context</summary>

Already covered by existing review threads, so not repeated here:
- The updater-vs-atom architecture feedback on `BalancesWatcherUpdater` / `NativeTokenBalanceUpdater`.
- The `toLowerCase()` test normalization nit, which is already resolved with `getAddressKey`.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

```text
Verify this finding against current PR head. Keep the fix minimal and validate it with focused tests.

Context:
- `CommonPriorityBalancesAndAllowancesUpdater` switches EVM chains to `BalancesWatcherUpdater` when `isBwEnabled` is true.
- `useBalancesWatcherSession` sets `error` and `isLoading: false` on create-session rejection or terminal SSE error, but leaves `hasFirstLoad: false`.
- `useTradeFormValidationContext` computes `isBalancesLoading` as `!hasFirstLoad || isBalancesLoading`, so pre-snapshot watcher failures remain stuck as loading.
- Expected behavior: pre-snapshot watcher failures should render a balance-load error/not-loaded state instead of permanent loading.
```

</details>

_Generated using the `pr-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>

**Comment by fairlighteth** [comment]
<details>
<summary>🧪 <strong>Scoped browser QA passed</strong>: watcher session + visible balance rows; swap execution not covered</summary>

AI usage: Codex drove the Playwright run and drafted this summary from the listed scenarios and QA artifacts.

**Outcome**
- Watcher session setup passed in Chromium and Firefox: `POST /sessions` returned `200`, and the SSE endpoint opened with `200`.
- Enabled-list token balances were visible in the swap form and token selector.
- Custom-token session refresh was verified at the network layer: a later `/sessions` request included `customTokens`.
- Not covered: real order execution, signing-capable wallet flow, and 0-2s post-execution balance update timing.

**Run details**
- Source: Cloudflare PR preview `https://42ec8792.swap-dev-5u6.pages.dev`
- Commit: `67445bd9194f1ce76cb7a571a8c4a0d1b236e986`
- Browsers: Chromium `140.0.7339.16`, Firefox `141.0`
- OS: macOS
- Wallet mode: `wallet-connected-ui` with mocked EIP-1193/EIP-6963 provider
- Mocked address: `0xCa011D01A4B75a36AFB5e2B69E0BA8F01c6B500E`
- Local worktree was not at the PR head, so the deployed preview matching the PR commit was used intentionally.

**How to retest**
1. Open the PR preview on Ethereum with `isBwEnabled` enabled.
2. Connect a wallet or mocked provider.
3. Click through the swap promo card if shown.
4. Confirm selected-token balances are visible in the swap form.
5. Open the token selector and confirm enabled-list token balances are visible.
6. Add/search a custom token and confirm a later `/sessions` request includes `customTokens`.

**Artifacts**
- Chromium watcher/session recording: https://github.com/cowprotocol/cowswap/releases/download/pr-qa-assets/prqa-7640-chromium-wallet-connected-ui-watcher-session.mp4
- Firefox watcher/session recording: https://github.com/cowprotocol/cowswap/releases/download/pr-qa-assets/prqa-7640-firefox-wallet-connected-ui-watcher-session.mp4
- Chromium balance walkthrough: https://github.com/cowprotocol/cowswap/releases/download/pr-qa-assets/prqa-7640-chromium-balance-visual-walkthrough.mp4
- Firefox balance walkthrough: https://github.com/cowprotocol/cowswap/releases/download/pr-qa-assets/prqa-7640-firefox-balance-visual-walkthrough.mp4

Chromium token-selector balances:

Chromium token selector balances visible

Firefox token-selector balances:

Firefox token selector balances visible

**Not checked / follow-up**
- Real order execution / signing-capable wallet flow.
- 0-2s post-execution balance update timing.
- Visual custom/imported-token balance update after adding a custom token; only the custom-token session refresh was verified at the network layer.
</details>

File: libs/balances-and-allowances/src/updaters/BalancesWatcherUpdater.tsx (hunk 1 above)
Comment (limitofzero) [reply libs/balances-and-allowances/src/updaters/BalancesWatcherUpdater.tsx:30]: get it, thanks, will rewrite this part

Comment (limitofzero) [reply libs/balances-and-allowances/src/updaters/BalancesWatcherUpdater.tsx:30]: hmm, I've just checked the scope of work, and it requires more changes that I expected. 
First I need to move feature flags from Context to atoms, the same for hooks - I can't use them in onMount level, so I'm not sure that it's a good start here regarding the huge task scope

**Review by elena-zh — COMMENTED** [review]
Hey @limitofzero , thank you, works really great!!

Some nitpicks: 
1. I cannot load balances for some tokens on Base and BNB 
 - pick a sell token on ehterreum
 - open buy token selector on BNB or BASE

Image
Image

2. Is it OK that I see 3 bw post requests  on Account-overview page/other pages 

Image

3. Is it OK that I don't see bw requests on console when an order is filled? Balances are updated, just there is no bw request on console

4. I know that it should not be there, but it would be great to have the same feature for allowances update ;) 

Thank you!

**Comment by limitofzero** [comment]
> Hey @limitofzero , thank you, works really great!!


1. @elena-zh , thank you for your work! I started to check your first discovered issue and I have the on prod with multicall.
image 
So I it's a prod behavior and I suggest to not block this pr and investigate [the reason](https://linear.app/cowswap/issue/FE-203/some-token-balances-are-not-fetching-via-multicall)

2. nice catch, it's not ok, but it's an issue of our current architecture, I'm not going to address it in that pr(because it can affect a lot of our app parts. Currently it doesn't affect the frontend, it only creates some load on BW service. Will address it later! I've created a [bug](https://linear.app/cowswap/issue/FE-202/few-session-request-on-account-page)

3. there is another endpoint that u can find by adding sse text in filter. It's should be only one request, initially it returns the balance of all requested tokens, then only diffs. 
telegram-cloud-photo-size-2-5305482382909577754-y

4. yep, it's our next goal. First we want to test current load on balances and adjust balances-watcher to handle more connections and ws subscriptions

**Review by elena-zh — APPROVED** [review]
Thank you, approved with issues to be addressed separately!

**Comment by limitofzero** [comment]
> **AI Review** (Codex GPT-5, worked 5m): watcher failures can leave swaps stuck on balance loading
> **Finding:** Handle watcher errors before the first snapshot as a completed balance load failure
> 
> * Location: `libs/balances-and-allowances/src/hooks/useBalancesWatcherSession.ts:75` and `apps/cowswap-frontend/src/modules/tradeFormValidation/hooks/useTradeFormValidationContext.ts:111`
> * When `isBwEnabled` is true, the PR bypasses the RPC updater and relies on the watcher path for EVM chains.
> * If `createBalancesWatcherSession` rejects or the SSE stream terminates before the first `balance_update`, the hook sets `error` and `isLoading: false`, but leaves `hasFirstLoad: false`.
> * The validation context then computes `isBalancesLoading: !hasFirstLoad || isBalancesLoading`, so the UI stays in `BalancesLoading` instead of surfacing `BalancesNotLoaded`. This matches the current Cypress failures on the PR head where swap flows never find `#do-trade-button`.
> 
> **Suggested fix**
> 
> * Make the error path stop the first-load gate, or make validation treat a balance error as not loading. For example, terminal watcher/create-session failures should transition to a state that renders `BalancesNotLoaded` rather than permanent `BalancesLoading`.
> * Add coverage for both `createSession` rejection and terminal SSE error before the first snapshot, asserting the downstream validation sees a not-loaded/error state rather than loading forever.
> 
> Review scope and related context

fixed, thank you

**Review by fairlighteth — COMMENTED** [review]
<details>
<summary>⚠️ <strong>AI Review</strong> (Codex GPT-5, worked 4m): empty watcher token set can keep balances loading forever</summary>

**Finding:** Mark watcher mode loaded when there are no ERC20 tokens to subscribe to

- Location: `libs/balances-and-allowances/src/hooks/useBalancesWatcherSession.ts:43`
- When `isBwEnabled` is on, `BalancesWatcherUpdater` replaces the RPC updater for EVM chains.
- If a connected user has no enabled token-list URLs and no custom tokens, `useBalancesWatcherSession` returns before creating a session and never updates `balancesAtom`.
- `useTradeFormValidationContext` still treats `!hasFirstLoad` as loading, so the swap UI can stay in `BalancesLoading` forever even though `NativeTokenBalanceUpdater` may separately write the native balance.

**Suggested fix**
- In the empty `tokensListsUrls` + `customTokens` case, transition `balancesAtom` to a completed empty-load state for the current `chainId` (`hasFirstLoad: true`, `isLoading: false`, `error: null`) instead of returning without touching state.
- Add a focused test for a connected EVM account with both arrays empty, asserting the watcher does not create a session and the first-load gate is closed.

<details>
<summary>Review scope and related context</summary>

Already covered or rechecked, so not repeated as new findings:
- The previous AI finding about pre-snapshot watcher failures is fixed on the current head: terminal SSE errors and `createSession` rejection now call `applyTerminalError`, which sets `hasFirstLoad: true`.
- The `toLowerCase()` test normalization issue is resolved with `getAddressKey`.
- The updater-vs-atom architecture discussion is already active in Danziger's review threads.
- CodeRabbit already raised the non-blocking point that legacy hooks still run before the watcher-mode return.

Local targeted tests were not runnable in this checkout because dependencies are not installed (`node_modules` / Nx / Jest missing). GitHub Test/Lint/Typecheck/Cypress were still pending when checked.

</details>

<details>
<summary>🤖 Prompt for AI agents</summary>

```text
Verify this finding against the current PR head. Fix only if still valid, keep the change minimal, and validate with a focused test.

Context:
- `CommonPriorityBalancesAndAllowancesUpdater` renders `BalancesWatcherUpdater` for EVM chains when `isBwEnabled` is true.
- `useBalancesWatcherSession` returns early when `tokensListsUrls.length === 0 && customTokens.length === 0`.
- That early return does not set `hasFirstLoad: true`.
- `useTradeFormValidationContext` treats `!hasFirstLoad` as balances loading.
- Expected behavior: connected EVM watcher mode with no ERC20 tokens to subscribe to should complete an empty balances load instead of leaving the swap form stuck in loading.
```

</details>

_Generated using the `pr-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>

**Review by fairlighteth — COMMENTED** [review]
<details>
<summary>✅ <strong>Security Review</strong> (Codex GPT-5, worked 1m): no verified security findings</summary>

Review completed. I found no verified security findings worth posting.

<details>
<summary>Review scope and related context</summary>

- Checked: balances-watcher session creation, SSE URL construction, malformed `balance_update` handling, terminal/non-terminal error handling, stale session cleanup, token-list/custom-token data sent to the watcher, and workflow env wiring for `REACT_APP_BALANCES_WATCHER_BASE_URL`.
- Source: code inspection only, plus current GitHub PR checks.
- Lifecycle: no `verified-vulnerability` findings.
- Existing comments: the remaining issue I found earlier is availability/UX (`hasFirstLoad` stays false with an empty tracked token set), not a verified security vulnerability.
- CI context: lint, tests, typecheck, agent harness, Socket checks, and deploy previews pass on current head; Cypress was still pending when checked.
- Confidence limits: I did not inspect the balances-watcher backend, so backend-side handling of user-enabled token-list URLs is outside this frontend PR review.

</details>

_Generated using the `security-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>

**Comment by fairlighteth** [comment]
@limitofzero one more new finding https://github.com/cowprotocol/cowswap/pull/7640#pullrequestreview-4531597056 please review if applicable or not.

**Comment by limitofzero** [comment]
> @limitofzero one more new finding [#7640 (review)](https://github.com/cowprotocol/cowswap/pull/7640#pullrequestreview-4531597056) please review if applicable or not.

thank you, fixed

**Review by fairlighteth — COMMENTED** [review]
<details>
<summary>✅ <strong>AI Review</strong> (Codex GPT-5, worked 1m): follow-up addressed</summary>

**Rechecked**
- Author response: @limitofzero said the empty watcher token-set finding was fixed.
- Code path: `libs/balances-and-allowances/src/hooks/useBalancesWatcherSession.ts:43`.
- Test coverage: `libs/balances-and-allowances/src/hooks/useBalancesWatcherSession.test.tsx:110`.

**Result:** Fixed. The empty token-list/custom-token case now skips session creation but still closes the first-load gate with `hasFirstLoad: true`, `isLoading: false`, `error: null`, and the current `chainId`.

<details>
<summary>🤖 Verification notes for AI agents</summary>

```text
Verify the prior finding against current code only. Confirm whether the author's claimed fix addresses the specific failure mode, and avoid reopening broader or unrelated issues.

Prior finding:
- Connected EVM watcher mode with no enabled token-list URLs and no custom tokens returned before updating balancesAtom, leaving useTradeFormValidationContext stuck in BalancesLoading.

Current verification:
- useBalancesWatcherSession now calls applyEmptyLoad(state, chainId) before returning from the empty arrays branch.
- The focused test asserts no session is created and hasFirstLoad is true, isLoading is false, error is null, and chainId is set.
```

</details>

_Generated using the `pr-review` skill from the [CoW Protocol skills repo](https://github.com/cowprotocol/skills)._

</details>

**Review by fairlighteth — APPROVED** [review]
Image
